Feature - Better error and exception handling

Rican7 opened the pull request:
This PR improves how Paulus 2 handles exceptions (and low-level errors)... by a lot. Here's how:
- The exception handling code has been abstracted into its own class, instead of being another giant method in the main `Paulus` class.
- Exception handlers are now settable/injectable into the application and are abstracted into generic interfaces, so that the exception handling behavior can be changed very easily. This allows us to handle exceptions in a number of different ways depending on environments, so for example you could print a colorful, interactive, HTML page when a RuntimeException is thrown in dev and silence all other exceptions in production.
- The default implementations of the exception handlers are both informative and BC compatible, giving the developer more information while keeping backwards compatibility
- The entire new structure is based on a dependency injection principle, making testing a lot easier. Now you can even provide a different exception handler when testing! Woot!
- Finally, we now support low-level native PHP error handling. **Finally**
- The new _error_ handler works in a very similar way to the exception handler, being a new generic interface-based injectable implementation into the main Paulus application class.

So yea, I've had this idea for quite a while now and finally had the niche of time to realize it. I hope everyone enjoys the new error handling.
#### Example

Here's an example of the new error handling in a verbose/informative format:

``` json
{
  "meta": {
    "status_code": 500,
    "status": "EXCEPTION_THROWN",
    "message": "sick",
    "more_info": {
      "exception_class": "ErrorException",
      "stack_trace": [
        {
          "function": "handleError",
          "class": "Paulus\\Handler\\Error\\BasicErrorHandler",
          "type": "->",
          "args": [
            16384,
            "sick",
            "/srv/www/dev/rican7/src/Api/Api.php",
            259,
            {
              "request": {},
              "response": null,
              "send_response": true,
              "capture": 0
            }
          ]
        },
        {
          "file": "/srv/www/dev/rican7/src/Api/Api.php",
          "line": 259,
          "function": "trigger_error",
          "args": [
            "sick",
            16384
          ]
        },
        {
          "file": "/srv/www/dev/rican7/index.php",
          "line": 13,
          "function": "run",
          "class": "Api\\Api",
          "type": "->",
          "args": []
        }
      ]
    }
  },
  "data": {}
}
```

Rican7 committed:
New interface for handling application exceptions

Rican7 committed:
New property, getter, and setter in the Paulus
class for the new ExceptionHandlerInterface

Rican7 committed:
Two new exception handler classes for Paulus'
default implementation of the ExceptionHandlerInterface

Rican7 committed:
Now using our RestfulExceptionHandler to handle
all exceptions in Paulus by default

Rican7 committed:
Removing the old exception handling code that was
directly in the Paulus class

Rican7 committed:
Our abstract controller now more properly rethrows
exceptions that it doesn't handle itself

Rican7 committed:
Removing unused use statements

Rican7 committed:
Reducing duplication between the two exception handlers

Rican7 committed:
New interface for exception handlers that use a
response object

Rican7 committed:
Abstracting out the default exception handler
implementations a bit more an giving the user more options

Rican7 committed:
Moving things around and making our default response
no longer a lazy getter

Rican7 committed:
New interface for error handlers

Rican7 committed:
Giving Paulus the ability to handle low-level
errors using the ErrorHandlerInterface

Rican7 committed:
Adding a new basic implementation of the
ErrorHandlerInterface and setting Paulus to use it

Rican7 committed:
Boom. Smarter error handling

Rican7 committed:
More information! Yay!

Rican7 committed:
Unused use

brianmuse commented on src/Paulus/Handler/Error/ErrorHandlerInterface.php:
```diff
@@ -0,0 +1,38 @@
+     * @return boolean
+     */
+    public function handleError($level, $message, $file = null, $line = null, array $context = null);
+}
```
Would be helpful to document the params here...specifically `$context` might need some details

Rican7 replied:
Ah, yea, good point. Especially since its an interface. :+1:

brianmuse commented on src/Paulus/Handler/Exception/BasicExceptionHandler.php:
```diff
@@ -0,0 +1,137 @@
+        // Send our response
+        $this->response->send();
+
+        return true;
```
Is `false` ever returned? Should this just return void?

Rican7 replied:
It isn't in these cases, but I didn't want to violate the interface here and I built the interface for future compatibility with a feature of chaining handlers so if one returns false, it moves to the next handler. That feature isn't dire right now and would require its own PR, but I wanted the interface to be future-compatible.

brianmuse commented on src/Paulus/Handler/Exception/RestfulExceptionHandler.php:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Handle an exception
+     *
+     * Handle an exception thrown in the application
```
Being nitpicky, but pretty redundant. Can probably just be the second line.

Rican7 replied:
Agreed

jstruzik commented on src/Paulus/Handler/Exception/BasicExceptionHandler.php:
```diff
@@ -0,0 +1,137 @@
+     */
+    public function setResponse(AbstractResponse $response)
+    {
+        $this->response = $response;
```
You say `@return BasicExceptionHandler`, but there's no `return $this;`

Rican7 replied:
Goooood catch. Ty

brianmuse commented on src/Paulus/Handler/Exception/RestfulExceptionHandler.php:
```diff
@@ -0,0 +1,150 @@
+            return $this->handleRestfulException($exception);
+
+        } elseif (null !== $this->delegate) {
+            return $this->delegate->handleException($exception);
```
I really like this! :+1:

Rican7 replied:
Haha, thanks. I realized that we'll probably want to use the default RestfulExceptionHandler in our app, but not exactly the parent behaviors, so that's why I did this. Nice feature to have anyway. :P

brianmuse commented on src/Paulus/Paulus.php:
```diff
@@ -56,7 +58,7 @@ class Paulus
      * @const string
      */
-    const FALLBACK_RESPONSE_CLASS = '\Klein\Response';
+    const DEFAULT_RESPONSE_CLASS = '\Klein\Response';
```
Is it really a "default" if it's used exclusively?

Rican7 replied:
Its just the one that's used by default (so that the response class isn't null), but its designed to be overwritten: https://github.com/onemightyroar/PHP-Paulus-Components/blob/a25bdcf7a5e1f98194f93296f982617fdd296f9d/src/OneMightyRoar/PHP_Paulus_Components/ApiApp.php#L60-L62

brianmuse replied:
Ah ok, I was reading it as the "default" "response_class", but really its the "default_response" "class" haha

jstruzik commented on src/Paulus/Handler/Exception/InformativeExceptionHandler.php:
```diff
@@ -0,0 +1,128 @@
+     * @return void
+     */
+    protected function prepareErrorResponse(
+        $code = self::DEFAULT_RESPONSE_CODE,
```
Why are you setting this in the parameters when you handle it at the start of the method?

Rican7 replied:
That way the IDE/tooling documents the defaults (since they're static, not object instances, so we can do this here)

jstruzik commented on src/Paulus/Handler/Exception/InformativeExceptionHandler.php:
```diff
@@ -0,0 +1,128 @@
+                ->setMessage($message);
+
+            if (null !== $more_info) {
+                $response->setMoreInfo($more_info);
```
Even if `$more_info` were null here, you're going to get the same result [here](https://github.com/Rican7/Paulus/blob/release/2.0.0/src/Paulus/Response/ApiResponse.php#L226).

So I don't think this condition is needed, just set it and forget it!

Rican7 replied:
I don't want to override the `more_info` state that might already exist in the set exception.

jstruzik replied:
Well played!

jstruzik commented on src/Paulus/Handler/Exception/RestfulExceptionHandler.php:
```diff
@@ -0,0 +1,150 @@
+     * @access public
+     * @return boolean
+     */
+    public function handleRestfulException(ApiExceptionInterface $exception)
```
Very nice!

jstruzik commented:
:+1: Very good addition!

Rican7 committed:
Doc-blocks updated to be more informative and less
redundant

mcos commented on src/Paulus/Handler/Error/BasicErrorHandler.php:
```diff
@@ -0,0 +1,49 @@
+    public function handleError($level, $message, $file = null, $line = null, array $context = null)
+    {
+        // If the error level has been disabled in our error reporting configuration
+        if (!($level & error_reporting())) {
```
Shouldn't this be `&&`, or are you trying to do perform a bitwise operation?

Rican7 replied:
Doing a bitwise here. We want to see if the error level is something that we've configured PHP to report or not.

mcos replied:
Gotcha. Thanks. Was a little confused since it's unusual (and a bit dangerous) to see a bitwise operation in a conditional.

Rican7 committed:
Whoooops. Returning what we said we'd return

Rican7 committed:
Now splitting up the BasicErrorHandler and creating
a new LevelBasedErrorHandler that calls the basic error handler. That
way we won't have to reproduce the ErrorException code for other
handlers and it allows for an application to easily just ignore the
`error_reporting` directive if they so choose.

Rican7 committed:
Now using the LevelBasedErrorHandler by default

Rican7 commented:
@brianmuse @jstruzik @mcos 
Ok guys. Made some updates. :)

brianmuse commented:
:+1:!

jstruzik commented:
:+1:
